corehq/util/couch_doc_processor.py

```diff
@@ -0,0 +1,369 @@
+    before giving up.
+    :param chunk_size: Maximum number of records to read from couch at
+    one time. It may be necessary to use a smaller chunk size if the
+    records being processed are very large and the default chunk size of
+    100 would exceed available memory.
+    """
+    def __init__(self, doc_type_map, doc_processor, reset=False, max_retry=2, chunk_size=100):
+        self.doc_type_map = doc_type_map
+        self.doc_processor = doc_processor
+        self.reset = reset
+        self.max_retry = max_retry
+        self.chunk_size = chunk_size
+
+        self.couchdb = next(iter(doc_type_map.values())).get_db()
+        assert all(m.get_db() is self.couchdb for m in doc_type_map.values()), \
+            "documents must live in same couch db: %s" % repr(doc_type_map)
+
+        self.docs_by_type = ResumableDocsByTypeIterator(
+            self.couchdb, doc_type_map, doc_processor.unique_key, chunk_size=chunk_size
+        )
+
+        self.visited = 0
+        self.previously_visited = 0
+        self.total = 0
+
+        self.processed = 0
+        self.skipped = 0
+
+        self.start = None
+
+    def has_started(self):
+        return bool(self.docs_by_type.progress_info)
+
+    @property
+    def session_visited(self):
+        return self.visited - self.previously_visited
+
+    @property
+    def session_total(self):
+        return self.total - self.previously_visited
+
+    @property
+    def attempted(self):
+        return self.processed + self.skipped
+
+    @property
+    def timing(self):
+        """Returns a tuple of (elapsed, remaining)"""
+        elapsed = datetime.now() - self.start
+        if self.session_visited > self.session_total:
+            remaining = "?"
+        else:
+            session_remaining = self.session_total - self.session_visited
+            remaining = elapsed / self.session_visited * session_remaining
+        return elapsed, remaining
+
+    def _setup(self):
+        from corehq.dbaccessors.couchapps.all_docs import get_doc_count_by_type
+
+        self.total = sum(get_doc_count_by_type(self.couchdb, doc_type)
+                    for doc_type in self.doc_type_map)
+
+        if self.reset:
+            self.docs_by_type.discard_state()
+        elif self.docs_by_type.progress_info:
+            info = self.docs_by_type.progress_info
+            old_total = info["total"]
+            # Estimate already visited based on difference of old/new
+            # totals. The theory is that new or deleted records will be
+            # evenly distributed across the entire set.
+            self.visited = int(round(float(self.total) / old_total * info["visited"]))
+            self.previously_visited = self.visited
+        print("Processing {} documents{}: {}...".format(
+            self.total,
+            " (~{} already processed)".format(self.visited) if self.visited else "",
+            ", ".join(sorted(self.doc_type_map))
+        ))
```

Comment:
Now that this code has been converted from a command line utility into a library class, would it make sense to move progress output (print statements) into the doc processor? I assume the pillowtop reindexer code will not be run from the command line, so print statements don't seem to make sense there.